src/osu/HitObject.java

```diff
@@ -153,119 +165,110 @@ public HitObject clone(){
 		}
 		ho.setSetID(setID);
-		ho.setAddition(addition);
+		ho.setAddition(additionSampleSet);
 		ho.setSampleSet(sampleSet);
 		ho.setTPSampleSet(timingPointSampleSet);
 		ho.setTimingPointVolume(timingPointVolume);
 		return ho;
 	}
 	
 	public int getAddition(){
-		return addition;
+		return additionSampleSet;
 	}
 	
 	public void addWhistleFinishClap(int value){
-			setWhislteFinishClap(whistle_finish_clap+value);
+			setWhislteFinishClap(additionFlags+value);
 	}
 	
 	public void addWhistleFinishClap(int value1, int value2){
-		setWhislteFinishClap(whistle_finish_clap+value1+value2);
+		setWhislteFinishClap(additionFlags+value1+value2);
 	}
 	
 	public void setAddition(int v){
-		addition=v;
+		additionSampleSet=v;
 	}
 	
 	public Sample toSample(){
 		Sample s;
-			if (!hitSound.contains(".wav")){
-				
+		if (!usesCustomSampleFile()) {
+			String sampleSound = hitSound;
+			int effectiveSampleSet;
 
-				
-				switch(timingPointSampleSet){
-				
-				case 1:
-					hitSound = "normal-";
-					break;
-					
-				case 2:
-					hitSound = "soft-";
-					break;
-					
-				case 3:
-					hitSound = "drum-";
+			if (additionFlags == SAMPLESET_AUTO)
+				effectiveSampleSet = timingPointSampleSet;
+			else
+				effectiveSampleSet = additionSampleSet;
+
+			switch (effectiveSampleSet){
+				case SAMPLESET_NORMAL:
+					sampleSound = "normal-";
 					break;
-				}
 				
-				if (whistle_finish_clap != HITNORMAL)
-				switch(addition){
-				case 1:
-					hitSound = "normal-";
+				case SAMPLESET_SOFT:
+					sampleSound = "soft-";
 					break;
-					
-				case 2:
-					hitSound = "soft-";
-					break;
-					
-				case 3:
-					hitSound = "drum-";
-					break;
-				}
-		
 				
-				switch(whistle_finish_clap){
+				case SAMPLESET_DRUM:
+					sampleSound = "drum-";
+					break;
+			}
+
+			switch (additionFlags){
 				case HITNORMAL:
-					hitSound+="hitnormal";
+					sampleSound+="hitnormal";
 					break;
 				case HITWHISTLE:
-					hitSound+="hitwhistle";
+					sampleSound+="hitwhistle";
 					break;
 				case HITFINISH:
-					hitSound+="hitfinish";
+					sampleSound+="hitfinish";
 					break;
 				case HITCLAP:
-					hitSound+="hitclap";
-				}
-				
-				String id = "";
-				if (setID>1){
-					id+=setID;
-				}
-				hitSound+=id+".wav";
-				s = new Sample(startTime,hitSound,timingPointVolume);
-				System.out.println(timingPointVolume);
-			} else {
-				s = new Sample(startTime,hitSound,volume);
+					sampleSound+="hitclap";
+					break;
 			}
-			
-			s.addQuotesToHS();
-			if (!s.toString().contains(".wav")){
-				System.err.println("Failed to convert HitObject to Sample");
-				System.err.println(toString());
-				System.err.println(s.toString());
-				System.exit(-1);
+
+			String id = "";
+
+			if (setID > 1) {
+				id += setID;
 			}
-			return s;
+
+			// Assume the normal hitsounds are wav files. This is
+			// the format in which DH & Co. provide their hitsounds.
+			sampleSound += id + ".wav";

```

Comment:
the problem here is that when you want make a WCF into a SB note in lower diff. you don't know whether it's wav or ogg. since the WFC hs dont specify whether its wav or ogg. and its possible to have both soft-hitclap.ogg and soft-hitclap.wav in the same folder both used in different places. 

I did some testing and it seems when both is present the .wav is used instead. pls do some testing on your side too just to confirm this. 

so here we need to do a look up in the folder to check if the sample exist for wav format, if not check if ogg exist. if neither exist then default to wav.

Reply:
I've assumed that `.wav` is always used as the extension name here, not considering other cases, as written in the comments. The behaviour of osu-stable (a black box) is relied on in case the wav file does not exist.